### ISSUE TITLE
MySQL: AsString generates VARCHAR too big for MySQL

### ISSUE THREAD

**Issue by escarls** [schambers/fluentmigrator#774]
After #691, the generated SQL is not always valid in MySQL.
The reason for this is that the legal max size of the VARCHAR column is not 65535 chars.
It's 65535 bytes (not chars) divided on all the columns in a raw.
So knowing this, the size of the column also depends on what character set that is defined for that table.
Also one have to count in the storage needed to store the size VARCHAR column, which is 1 byte for 1-255 chars and 2 bytes for 256-65535 (technically 65533 chars) if the character set is "latin 1".
However, if you change character set to UTF8, then the column size is reduced to 21845 characters, or if you set it to UTFMB4, then it's reduced to 16383 characters.

All of what I wrote above is still only valid if the VARCHAR column is alone on the line.
If you add an integer Id field, then you have 4 less bytes to work with, so for a "latin 1" charset table, then the VARCHAR max would be 65529.

As you can see, the max size of VARCHAR after the 255 max size was removed, is not a fixed one, it depends on what other fields you have in your table and what character set the table is.

I did see issue #517, but I do think that if you have a specific need to generate VARCHARS with size greater than 255, then maybe it's better to just do a AsCustom()?

So in conclusion, would it be possible to just revert pr #691, as it kind of breaks more than it fixes?

**Comment by fubar-coder** [comment]
What about a compromise? Using VARCHAR with a length of up to 8k bytes and using TEXT otherwise?

**Comment by escarls** [comment]
Changing it to 8k works a lot better than this.
I can cook up a PR for the changes if you want?

**Comment by fubar-coder** [comment]
Sure

**Comment by escarls** [comment]
With this patch all of my test works.
It's still possible to make FluentMigrator generate SQL that fails, but so it would have if VARCHAR size was set back to 255 also.
I'm closing the issue.
